### PR TITLE
Use appropriate provider for SHM DNS record

### LINKED ISSUE #2201: Unable to deploy SHM and SRE to different subscriptions
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** 
- **Data Safe Haven version:** v5.0.0

## :package: Packages

<details>
<summary>List of packages</summary>

```none
Paste list of packages here
```

</details>

## :no_entry_sign: Describe the problem

It is not possibly to deploy a TRE across subscriptions, with SREs deployed to subscriptions not containing their SHM.

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
Your log details here
```

</details>

## :recycle: To reproduce

- Deploy an SHM to SubscriptionA
- Attempt to deploy an SRE to SubscriptionB

### PR DESCRIPTION
**Pull request by JimMadge**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :closed_umbrella: Related issues

Closes #2201 

### :microscope: Tests

- Single subscription deployment ✅
- Multiple subscription deployment ✅

### CODE REVIEW

File: data_safe_haven/infrastructure/programs/sre/networking.py [hunk 1]
```diff
@@ -1856,6 +1858,13 @@ def __init__(
             opts=child_opts,
             tags=child_tags,
         )
+        shm_provider = provider.Provider(
+            "shm_provider",
+            provider.ProviderArgs(
+                location=props.location,
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/programs/sre/networking.py]: Is it always the case that the SHM and SRE will be deployed in the same region? Do we want/need this to be the case? If not, let's read the `shm_location` from the SHM config file.

Comment (JimMadge) [reply data_safe_haven/infrastructure/programs/sre/networking.py]: Added in d327acd

**Comment by JimMadge** [comment]
I'm getting a failure to create the SSL cert.

```console
 +  pulumi-python:dynamic:Resource sre_data_kvc_https_certificate
**creating failed** error: Exception calling application: Failed to
create SSL certificate resrudh-kernow-develop-turingsafehaven-ac-uk for
resrudh.kernow.develop.turingsafehaven.ac.uk.  Failed to create DNS TXT record _acme-challenge
in zone resrudh.kernow.develop.turingsafehaven.ac.uk.
```

Azure SDK says it failed to create the TXT record. Feels like an odd thing to crop up here as this is all within the SRE subscription. The DNS resource already has a set of records deployed with Pulumi.

Possibly related to #2209.

Any thoughts @jemrobinson.

**Comment by jemrobinson** [comment]
Is this perhaps using the SHM provider instead of the (default) SRE one?

**Comment by jemrobinson** [comment]
- The `Failed to create DNS TXT record` message comes from `AzureSdk::ensure_dns_txt_record`
- The `Failed to create SSL certificate` message comes from `SSLCertificateProvider::create` which calls `AzureSdk::ensure_dns_txt_record`

the entire relevant section of code is:

```python
client = ACMEClient(
    domains=[props["domain_name"]],
    email=props["admin_email_address"],
    directory="https://acme-v02.api.letsencrypt.org/directory",
    nameservers=["8.8.8.8", "1.1.1.1"],
    new_account=True,
)
private_key_bytes = client.generate_private_key(key_type="rsa2048")
client.generate_csr()
azure_sdk = AzureSdk(props["subscription_name"], disable_logging=True)
verification_tokens = client.request_verification_tokens().items()
for record_name, record_values in verification_tokens:
    record_set = azure_sdk.ensure_dns_txt_record(...)
```

which makes me feel like perhaps the `ACMEClient` is not being correctly created and/or the `generate_private_key()` or `generate_csr()` functions aren't doing what we expect. Can you retry with some manual logging interventions (e.g. set `disable_logging=False` in the AzureSdk call and also add some `logger.info` lines to help diagnose?)

**N.B.** this uses the production let's encrypt server, so for debugging it's worth manually changing to the staging server.

**Comment by JimMadge** [comment]
Yeah that's what I've been looking at. I'll do a bit more digging.

I would be surprised if it were a wrong subscription problem as the SRE subscription gets passed to the dynamic provider as an argument, then the Azure SDK client is created using that argument.

**Comment by JimMadge** [comment]
I can query the DNS zone using AZ CLI so the permissions should be correct.

**Comment by jemrobinson** [comment]
This could be the same SSL certificate problem as #2209, i.e. unrelated to these subscription changes.

**Comment by JimMadge** [comment]
On a positive note. Creating the NS records in the SHM DNS zone works :+1:.

**Comment by JimMadge** [comment]
Getting somewhere now,

```console
**creating failed** error: Exception calling application: Failed to
create SSL certificate porthperan-kernow-develop-turingsafehaven-ac-uk for
porthperan.kernow.develop.turingsafehaven.ac.uk.  Failed to create DNS TXT record
_acme-challenge in zone porthperan.kernow.develop.turingsafehaven.ac.uk.
(ResourceGroupNotFound) Resource group 'shm-kernow-sre-porthperan-rg' could not be found.
```

**Comment by JimMadge** [comment]
The AzureSDK class _is_ using the wrong subscription 🤔.

**Comment by JimMadge** [comment]
https://github.com/alan-turing-institute/data-safe-haven/blob/27e9dcac26c5ca2b738731d0030aa7625c150930/data_safe_haven/infrastructure/programs/declarative_sre.py#L195

There we go 😆

**Comment by JimMadge** [comment]
Deployment succeeds at a0ac911.
SRE provisioning manager throws a similar error.

**Review by jemrobinson — APPROVED** [review]
LGTM.

**Pull request merged**